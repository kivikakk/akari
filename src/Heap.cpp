// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#include <Akari.hpp>
#include <Console.hpp>
#include <Heap.hpp>
#include <string>

Heap::Heap(ptr_t start, ptr_t end, ptr_t max, u32 index_size, bool supervisor, bool readonly):
_index(reinterpret_cast<Entry *>(start), index_size, IndexSort),
_start(start), _end(end), _max(max), _used(0), _supervisor(supervisor), _readonly(readonly)
{
	ASSERT(start % 0x1000 == 0);
	ASSERT(end % 0x1000 == 0);
	ASSERT(max % 0x1000 == 0);

	start += sizeof(Entry) * index_size;
	
	if (start & 0xFFFF)
		start = (start & ~0xFFF) + 0x1000;

	_used += (start - _start);
	_index.insert(Entry(start, end - start, true));
	ASSERT(start < end);
}

void *Heap::alloc(u32 n) {
	s32 it = smallestHole(n);
	ASSERT(it >= 0);		// TODO: resize instead

	Entry hole = _index[it];
	_index.remove(it);

	ptr_t dataStart = hole.start;
	if (n < hole.size) {
		// the hole is now forward by `n', and less by `n'
		hole.size -= n;
		hole.start += n;
		_index.insert(hole);
	} else {
		// nothing: it was exactly the right size
	}

	_used += n;

	Entry data(dataStart, n, false);
	_index.insert(data);
	return reinterpret_cast<void *>(dataStart);
}

void *Heap::allocAligned(u32 n) {
	s32 it = smallestAlignedHole(n);
	ASSERT(it >= 0);		// TODO: resize instead

	Entry hole = _index[it];
	_index.remove(it);

	ptr_t dataStart = hole.start;

	if (dataStart & 0xFFF) {
		// this isn't aligned! 
		u32 oldSize = hole.size;

		dataStart = (dataStart & ~0xFFF) + 0x1000;
		hole.size = 0x1000 - (hole.start & 0xFFF);
		_index.insert(hole);

		hole.start = dataStart;
		hole.size = oldSize - hole.size;
	}

	// by here, `hole' is an un-inserted hole, up to the end of
	// the hole we originally had, but starting where our data
	// wants to start.

	if (n < hole.size) {
		hole.size -= n;
		hole.start += n;
		_index.insert(hole);
	} else {
		// right size (as in Alloc())
	}

	_used += n;

	Entry data(dataStart, n, false);
	_index.insert(data);
	return reinterpret_cast<void *>(dataStart);
}

s32 Heap::indexOfEntryStarting(ptr_t start) const {
	for (u32 i = 0; i < _index.size(); ++i) {
		if (_index[i].start == start)
			return i;
	}

	return -1;
}

s32 Heap::indexOfEntryEnding(ptr_t end) const {
	for (u32 i = 0; i < _index.size(); ++i) {
		if (_index[i].start + _index[i].size == end)
			return i;
	}

	return -1;
}

void Heap::free(void *p) {
	s32 i = indexOfEntryStarting((ptr_t)p);
	if (i == -1) return;

	_index[i].isHole = true;
	
	s32 b = indexOfEntryEnding((ptr_t)p);

	if (b != -1 && _index[b].isHole) {
		ptr_t new_size = _index[b].size + _index[i].size,
			  new_end = new_size + _index[b].start;

		_index[b].size = new_size;
		_index.remove(i);
		i = indexOfEntryEnding(new_end);
	}

	s32 e = indexOfEntryStarting(_index[i].start + _index[i].size);
	if (e != -1 && _index[e].isHole) {
		_index[i].size += _index[e].size;
		_index.remove(e);
	}
}

ptr_t Heap::start() const {
	return _start;
}

ptr_t Heap::used() const {
	return _used;
}

const OrderedArray<Heap::Entry> &Heap::index() const {
	return _index;
}

Heap::Entry::Entry(ptr_t start, u32 size, bool isHole):
start(start), size(size), isHole(isHole)
{ }

bool Heap::IndexSort(const Entry &a, const Entry &b) {
	return a.size < b.size;
}

s32 Heap::smallestHole(u32 n) const {
	u32 it = 0;
	ASSERT(_index.size() > 0);

	while (it < _index.size()) {
		const Entry &entry = _index[it];
		if (!entry.isHole) {
			++it; continue;
		}
		if (entry.size >= n) {
			return it;
		}
		++it;
	}
	Akari->console->printf("looking for %x bytes, couldn't find\n", n);
	AkariPanic("no smallest hole in SmallestHole!");
	return -1;
}

s32 Heap::smallestAlignedHole(u32 n) const {
	u32 it = 0;
	while (it < _index.size()) {
		const Entry &entry = _index[it];
		if (!entry.isHole) {
			++it; continue;
		}

		// calculate how far we have to travel from the start to get to a page-align
		s32 off = 0;
		if (entry.start & 0xFFF)
			off = 0x1000 - (entry.start & 0xFFF);

		// if the size of this hole is large enough for our needs, considering we have
		// `off' bytes of waste, then it's good
		if ((static_cast<s32>(entry.size) - off) >= static_cast<s32>(n))
			return it;

		++it;
	}
	Akari->console->printf("looking for %x bytes, couldn't find\n", n);
	AkariPanic("no smallest hole in SmallestAlignedHole!");
	return -1;
}

