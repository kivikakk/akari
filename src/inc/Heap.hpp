// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#ifndef __HEAP_HPP__
#define __HEAP_HPP__

#include <arch.hpp>
#include <OrderedArray.hpp>

class Heap {
public:
	class Entry {
	public:
		Entry(ptr_t, u32, bool);

		ptr_t start; u32 size;
		bool isHole;
	};

	Heap(ptr_t start, ptr_t end, ptr_t max, u32 index_size, bool supervisor, bool readonly);

	void *alloc(u32);
	void *allocAligned(u32);
	void free(void *);

	ptr_t start() const;
	ptr_t used() const;
	const OrderedArray<Entry> &index() const;

protected:
	static bool IndexSort(const Entry &, const Entry &);

	s32 indexOfEntryStarting(ptr_t start) const;
	s32 indexOfEntryEnding(ptr_t end) const;
	s32 smallestHole(u32) const;
	s32 smallestAlignedHole(u32) const;

	OrderedArray<Entry> _index;
	ptr_t _start, _end, _max;
	ptr_t _used;
	bool _supervisor, _readonly;
};

#endif

