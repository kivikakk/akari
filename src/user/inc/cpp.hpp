// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#ifndef __CPP_HPP
#define __CPP_HPP

extern "C" void __cxa_pure_virtual();
extern "C" int __cxa_atexit(void (*func)(void *), void *arg, void *dso_handle);
extern "C" void exit(int status);
extern "C" void start(int argc, char **argv);
void *operator new(size_t);
void *operator new[](size_t);
void *operator new(size_t, void *);
void operator delete(void *);
void operator delete[](void *);

#endif

