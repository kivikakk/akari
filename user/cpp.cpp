// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#include <UserCalls.hpp>
#include <cpp.hpp>

extern "C" void __cxa_pure_virtual() {
	panic("__cxa_pure_virtual called in usermode");
}

void *operator new(size_t n) {
	return malloc(n);
}

void *operator new[](size_t n) {
	return malloc(n);
}

void *operator new(size_t, void *p) {
	return p;
}

void operator delete(void *p) {
	return free(p);
}

void operator delete[](void *p) {
	return free(p);
}

