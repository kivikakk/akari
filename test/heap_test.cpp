// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#include "test_helper.hpp"

#include <cstdio>

#include "Heap.hpp"
#include "../Heap.cpp"

void report(const Heap *h) {
	const OrderedArray<Heap::Entry> &e = h->index();

	printf("Heap with %ld entries\n", e.size());
	for (u32 i = 0; i < e.size(); ++i) {
		printf("  %ld:\t%llx\t+ %lx\t(%s)\n",
				i, e[i].start - h->start(), e[i].size, e[i].isHole ? "hole" : "data");
	}
}

int main() {
	char *area = new char[0x100fff];
	ptr_t aligned = (ptr_t)area;
	if (aligned & 0xfff)
		aligned = (aligned & ~0xfff) + 0x1000;

	ptr_t end = aligned + 0x100000;

	Heap *h = new Heap(aligned, end, end, 128, false, true);

	ptr_t ptr = (ptr_t) h->alloc(0x700);
	printf("offset from start: %llx\n", ptr - aligned);

	ptr_t ptr2 = (ptr_t) h->alloc(0x700);
	printf("offset from start: %llx\n", ptr2 - aligned);

	ptr_t ptr3 = (ptr_t) h->allocAligned(0x2);
	printf("offset from start: %llx\n", ptr3 - aligned);

	ptr_t ptr4 = (ptr_t) h->alloc(0x100);
	printf("offset from start: %llx\n", ptr4 - aligned);

	ptr_t ptr5 = (ptr_t) h->alloc(0x100);
	printf("offset from start: %llx\n", ptr5 - aligned);

	ptr_t ptr6 = (ptr_t) h->alloc(0x100);
	printf("offset from start: %llx\n", ptr6 - aligned);

	h->free((void *)ptr3);

	ptr_t ptr7 = (ptr_t) h->alloc(0x3);
	printf("offset from start: %llx\n", ptr7 - aligned);

	ptr_t ptr8 = (ptr_t) h->alloc(0x2);
	printf("offset from start: %llx\n", ptr8 - aligned);

	ptr_t ptr9 = (ptr_t) h->alloc(0x2);
	printf("offset from start: %llx\n", ptr9 - aligned);

	ptr_t ptr10 = (ptr_t) h->alloc(0x1);
	printf("offset from start: %llx\n", ptr10 - aligned);

	report(h);

	h->free((void *)ptr7);
	h->free((void *)ptr9);
	h->free((void *)ptr10);

	report(h);

	ptr_t ptr11 = (ptr_t) h->alloc(0x6);
	printf("offset from start: %llx\n", ptr11 - aligned);

	report(h);

	return 0;
}
