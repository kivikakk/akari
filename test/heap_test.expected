offset from start: 1000
offset from start: 1700
offset from start: 2000
offset from start: 1e00
offset from start: 1f00
offset from start: 2002
offset from start: 2102
offset from start: 2000
offset from start: 2105
offset from start: 2107
Heap with 10 entries
  0:	2107	+ 1	(data)
  1:	2105	+ 2	(data)
  2:	2000	+ 2	(data)
  3:	2102	+ 3	(data)
  4:	2002	+ 100	(data)
  5:	1f00	+ 100	(data)
  6:	1e00	+ 100	(data)
  7:	1700	+ 700	(data)
  8:	1000	+ 700	(data)
  9:	2108	+ fdef8	(hole)
Heap with 7 entries
  0:	2000	+ 2	(data)
  1:	2102	+ fdefe	(hole)
  2:	2002	+ 100	(data)
  3:	1f00	+ 100	(data)
  4:	1e00	+ 100	(data)
  5:	1700	+ 700	(data)
  6:	1000	+ 700	(data)
offset from start: 2102
Heap with 8 entries
  0:	2000	+ 2	(data)
  1:	2102	+ 6	(data)
  2:	2002	+ 100	(data)
  3:	1f00	+ 100	(data)
  4:	1e00	+ 100	(data)
  5:	1700	+ 700	(data)
  6:	1000	+ 700	(data)
  7:	2108	+ fdef8	(hole)
